plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/AsyncUploadUtils.java

```diff
@@ -0,0 +1,460 @@
+ * compatible open source license.
+ */
+
+package org.opensearch.repositories.s3.async;
+
+import com.jcraft.jzlib.JZlib;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.ExceptionsHelper;
+import org.opensearch.common.StreamContext;
+import org.opensearch.common.blobstore.exception.CorruptFileException;
+import org.opensearch.common.blobstore.stream.write.WritePriority;
+import org.opensearch.common.io.InputStreamContainer;
+import org.opensearch.common.unit.ByteSizeUnit;
+import org.opensearch.common.util.ByteUtils;
+import org.opensearch.repositories.s3.io.InputStreamCRC32Container;
+import org.opensearch.repositories.s3.SocketAccess;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.http.HttpStatusCode;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.AbortMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.ChecksumAlgorithm;
+import software.amazon.awssdk.services.s3.model.CompleteMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.CompleteMultipartUploadResponse;
+import software.amazon.awssdk.services.s3.model.CompletedMultipartUpload;
+import software.amazon.awssdk.services.s3.model.CompletedPart;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadResponse;
+import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartResponse;
+import software.amazon.awssdk.utils.CompletableFutureUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+
+/**
+ * A helper class that automatically uses multipart upload based on the size of the source object
+ */
+public final class AsyncUploadUtils {
+    private static final Logger log = LogManager.getLogger(AsyncUploadUtils.class);
+    private final ExecutorService executorService;
+    private final ExecutorService priorityExecutorService;
+    private final long minimumPartSize;
+
+    /**
+     * The max number of parts on S3 side is 10,000
+     */
+    private static final long MAX_UPLOAD_PARTS = 10_000;
```

Comment:
not able to find relevant PR , so commenting here : 

Should we make `DEFAULT_MULTIPART_UPLOAD_MINIMUM_PART_SIZE` much smaller value that 16 MB ? Did we do some tests to find this number ?

Reply:
Also for `eventLoopThreads` , can we check number of processors and have appropriate defaults , lets say `max(1/8th of procs, 1)`

Reply:
This is how I propose we breakdown the class
Lets move S3AsyncClient to the constructor and make it `AsyncTransferManager`, such this interface could be used across all transfers and becomes a single interface.
Then abstract out all the part specific logic and move to a `AsyncPartTransferHandler` that encapsulates the logic to transfer a single part. The `AsyncTransferManager` then can encapsulate all parts as needed for uploads and downloads.